### ISSUE TITLE
kernel idea: make MAX_FUNC_EXPR_NESTING global, use it in read.c, and lower it fro 1024 to e.g. 32

### ISSUE THREAD

**Issue by fingolfin** [gap-system/gap#2206]
Right now, there are at least two places which limit how deep functions (resp. closures) can be nested in each other:
* in `code.c`, we `#define MAX_FUNC_EXPR_NESTING 1024` and use that as a hard limit on how deep one can nest functions. Going beyond this limit triggers an assertion (resp. silently corrupts memory, if assertions are turned off).
* in `read.c`, specifically in `ReadCallVarAss`, we limit the nesting to 65536 on 32bit platforms and not at all (resp. to 2^48) on 64bit platforms:
```C
#ifndef SYS_IS_64_BIT
        // we store nest and indx together in var; if we are on a 32bit
        // system, then this limits both to 16 bits, which we enforce here
        if (nest >= 65536) {
            Pr("Warning: abandoning search for %s at 65536th higher frame\n",
               (Int)STATE(Value), 0L);
            break;
        }
#endif
```

These two limits should be aligned, say by making `MAX_FUNC_EXPR_NESTING` global. For technical reasons, it should stay a power of 2.

Furthermore,, I don't think I ever saw a nesting deeper than 4 or so (perhaps a bit more with the use of nested `x -> foo` lambda functions). Realistically, going much deeper becomes rather painful, and I doubt it matters. So we might also want to consider reducing the nesting limit to, say, 32. This frees up valuable bits which can be used to store variable reference more compactly (I am thinking about unifying the var identifiers for gvars, lvars, hvars, dvars into a single 32 / 64 bit word)... Anyway, right now there is no pressing reason to reduce the limit, but it will be easier if the code was refactored to use `MAX_FUNC_EXPR_NESTING` uniformly, which to me is another reason why this refactoring would be useful.

Lastly, I think it's not good that `code.c` crashes if the nesting is too deep -- it really should produce an error. To test a hypothetical fix for this, I'd temporarily reduce `MAX_FUNC_EXPR_NESTING` to a much smaller value, making it easier to trigger the problem with test code... Alternatively, one can write a function which generates a string containing code for a set of functions nested to a certain fixed level, in order to generate tests cases, even for the 1024 level function.

So in summary:
* make `MAX_FUNC_EXPR_NESTING` global and change `read.c` to use it,
* change `code.c` to give an error instead of crashing when `MAX_FUNC_EXPR_NESTING` is exceeded,
* add a test case for the latter,
* perhaps reduce `MAX_FUNC_EXPR_NESTING` to a smaller value

**Comment by fingolfin** [comment]
Also, in `read.c` we limit the number of function arguments resp. local vars (the latter contains the arguments as a subset) to 65536; and in `ReadCallVarAss`, we also implicitly assume this, when we squeeze it into 16 bits, e.g. like this
```C
        indx = findValueInNams(nams, 1, LEN_PLIST(nams));
        if (indx != 0) {
            type = (nest == 0) ? 'l' : 'h';
            var = (nest << 16) + indx;
            break;
        }
```
So this value 65536 resp. the number of bits should also be put into a global constant. And perhaps should also be lowered, say to 4096 (exceeding it already triggers a warning; we may wish to add an explicit test case for it. I have one around, though for 65536 vars, GAP gets super slower, due to the linear lookup on a plist `nams` of length `N`, which we perform `O(N)` times, for quadratic `O(N^2)` performance, which IIRC can translate into a few minutes for parsing such function.

**Comment by ChrisJefferson** [comment]
Reducing MAX_FUNC_EXPR_NESTING to 32 seems entirely reasonable.

It's not impossible that someone might have many local variables in a function, but 4096 also seems like a reasonable limit. Someone can always make an array if they want more.

**Comment by fingolfin** [comment]
Now that PR #2207 is submitted, I can clarify why I am interested in reducing those limits: doing so would allow me to squeeze `LHSRef` from PR #2207 into a single 32bit word, which would make it very convenient to store. That said, I can definitely squeeze it into 64bit anyway, which should be good enough.
